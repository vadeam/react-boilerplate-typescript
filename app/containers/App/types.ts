import { Record } from 'immutable'

import { Repo } from 'containers/RepoListItem/types'

export interface AppActionTypes {
  LOAD_REPOS: string
  REPOS_LOADED: string
  REPO_LOADING_ERROR: string
}

export interface LoadReposAction extends Action<AppActionTypes> {}

export interface LoadReposSuccessAction extends Action<AppActionTypes> {
  readonly repos: Nullable<Repo[]>
  readonly username: string
}

export interface LoadReposErrorAction<T = object> extends Action<AppActionTypes> {
  readonly error: T
}

export type AppActions = LoadReposAction | LoadReposSuccessAction | LoadReposErrorAction

export interface AppActionCreators {
  loadRepos: () => LoadReposAction
  reposLoaded: (repos: Nullable<Repo[]>, username: string) => LoadReposSuccessAction
  repoLoadingError: (error: object) => LoadReposErrorAction
}

export interface UserData {
  readonly repos: Nullable<Repo[]>
}

export const UserDataRecord = Record<UserData>({ repos: null }, 'UserDataRecord')

export interface AppState extends InheritedReducerState {
  readonly currentUser: string
  readonly userData: UserData
}

export const AppStateRecord = Record<AppState>(
  {
    isFetching: false,
    currentUser: '',
    userData: UserDataRecord(),
    error: null,
  },
  'AppStateRecord',
)
