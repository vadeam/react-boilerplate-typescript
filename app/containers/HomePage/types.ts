import { Record } from 'immutable'

export interface HomeActionTypes {
  CHANGE_USERNAME: string
}

export interface HomeAction extends Action<HomeActionTypes> {
  readonly username: string
}

export interface HomeActionCreators {
  changeUsername: (username: string) => HomeAction
}

export interface HomeState extends InheritedReducerState {
  readonly username: string
}

export const HomeStateRecord = Record<HomeState>(
  {
    username: '',
  },
  'HomeStateRecord',
)
