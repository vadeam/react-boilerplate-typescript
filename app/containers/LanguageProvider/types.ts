import { Record } from 'immutable'

import { LOCALE_EN, LOCALE_DE, DEFAULT_LOCALE } from 'constants/locales'

declare global {
  type Locales = typeof LOCALE_EN | typeof LOCALE_DE
}

export type LocalMessages = { [p: string]: string }

export interface LanguageActionTypes {
  CHANGE_LOCALE: string
  SET_LOCALE_MESSAGES: string
}

export interface ChangeLocaleAction extends Action<LanguageActionTypes> {
  readonly locale: Locales
}

export interface SetLocaleMessagesAction extends Action<LanguageActionTypes> {
  readonly messages: LocalMessages
}

export interface LanguageActionCreators {
  changeLocale: (locale: Locales) => ChangeLocaleAction
  setLocaleMessages: (messages: LocalMessages) => SetLocaleMessagesAction
}

export interface LanguageProviderState extends InheritedReducerState {
  readonly locale: Locales
  readonly messages: LocalMessages
}

export const LanguageProviderStateRecord = Record<LanguageProviderState>(
  {
    locale: DEFAULT_LOCALE,
    messages: {},
  },
  'LanguageProviderStateRecord',
)
