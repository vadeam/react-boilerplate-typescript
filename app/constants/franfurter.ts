export const DEFAULT_BASE_CURRENCY = 'EUR'
