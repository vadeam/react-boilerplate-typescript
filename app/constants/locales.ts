export const LOCALE_DE = 'de'
export const LOCALE_EN = 'en'
export const DEFAULT_LOCALE = LOCALE_EN

export const appLocales: Locales[] = [LOCALE_EN, LOCALE_DE]
export const CUSTOM_LOCALE_STORAGE_KEY = 'custom-locale'
