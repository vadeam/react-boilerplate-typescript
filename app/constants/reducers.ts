export const REDUCER_KEY_APP = 'app'
export const REDUCER_KEY_LANGUAGE = 'language'
export const REDUCER_KEY_ROUTER = 'router'
export const REDUCER_KEY_HOME = 'home'
// for testing purposes
export const REDUCER_KEY_TEST = 'test'
