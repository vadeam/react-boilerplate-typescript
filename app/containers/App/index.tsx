/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import * as React from 'react'
import { Helmet } from 'react-helmet-async'
import { hot } from 'react-hot-loader/root'
import { Route, Switch } from 'react-router-dom'

import Footer from 'components/Footer'
import Header from 'components/Header'
import FeaturePage from 'containers/FeaturePage/Loadable'
import HomePage from 'containers/HomePage/Loadable'
import NotFoundPage from 'containers/NotFoundPage/Loadable'
import { ThemeProvider } from 'styles/styled-components'
import defaultAppTheme from 'styles/default-theme'
import GlobalStyle from 'global-styles'

import { AppWrapperStyled } from './styled'

function App() {
  return (
    <ThemeProvider theme={defaultAppTheme}>
      <AppWrapperStyled>
        <Helmet titleTemplate="%s | React Boilerplate" defaultTitle="React Boilerplate">
          <meta name="description" content="React Boilerplate application" />
        </Helmet>
        <Header />
        <Switch>
          <Route exact path="/" component={HomePage} />
          <Route path="/features" component={FeaturePage} />
          <Route component={NotFoundPage} />
        </Switch>
        <Footer />
        <GlobalStyle />
      </AppWrapperStyled>
    </ThemeProvider>
  )
}

export default hot(App)
