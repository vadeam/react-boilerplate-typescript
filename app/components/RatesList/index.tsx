import * as React from 'react'
import useSWR from 'swr'

import List from 'components/List'
import ListItem from 'components/ListItem'
import RatesListItem, { RateItem } from 'containers/RatesListItem'
import fetcher from 'utils/fetcher'
import { DEFAULT_BASE_CURRENCY } from 'constants/franfurter'
import swrLaggy from 'utils/swrLaggy'

type RatesListProps = { baseCurrency?: string }
type RatesResponse = {
  amount: number
  base: string
  date: string
  rates: { [k: string]: number }
}

function RatesList({ baseCurrency = DEFAULT_BASE_CURRENCY }: RatesListProps) {
  const { data, error } = useSWR<RatesResponse>(`/latest?from=${baseCurrency}`, fetcher.getFranfurterData, {
    use: [swrLaggy],
  })

  if (error) {
    return <List component={() => <ListItem item="Something went wrong, please try again!" />} />
  }

  let rates: RateItem[] = []

  if (data?.rates) {
    rates = Object.keys(data.rates).map((key) => ({
      id: key,
      currency: key,
      rate: data.rates[key],
    }))
  }

  if (Array.isArray(rates) && rates.length > 0) {
    return <List<RateItem> items={rates} component={RatesListItem} />
  }

  return null
}

export default RatesList
