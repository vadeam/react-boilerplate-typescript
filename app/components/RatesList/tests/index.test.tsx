import { render, act } from '@testing-library/react'
import React from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'

import { DEFAULT_BASE_CURRENCY } from 'constants/franfurter'
import fetcher from 'utils/fetcher'
import history from 'utils/history'
import { AppStateRecord } from 'containers/App/types'
import { REDUCER_KEY_APP } from 'constants/reducers'
import { DEFAULT_LOCALE } from 'constants/locales'

import configureStore from '../../../configureStore'
import RatesList from '../index'

jest.mock('utils/fetcher')

describe('<RatesList />', () => {
  let fetcherMock: any

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should render an error if loading failed', async () => {
    fetcherMock = jest.spyOn(fetcher, 'getFranfurterData').mockImplementation(() => {
      throw new Error('Something went wrong')
    })

    const { queryByText } = await render(<RatesList />)

    expect(fetcherMock).toHaveBeenCalledTimes(1)
    expect(fetcherMock).toHaveBeenCalledWith(`/latest?from=${DEFAULT_BASE_CURRENCY}`)
    expect(queryByText(/Something went wrong/)).toBeInTheDocument()
  })

  it('should render the currencies rates if loading was successful', async () => {
    const baseCurrency = 'USD'
    const testRatesResponse = {
      amount: 1.0,
      base: baseCurrency,
      date: '2021-09-03',
      rates: {
        AUD: 1.3444,
        BGN: 1.6474,
        BRL: 5.1616,
      },
    }
    let renderResult: any = {}

    fetcherMock = jest.spyOn(fetcher, 'getFranfurterData').mockImplementation(() => Promise.resolve(testRatesResponse))

    await act(async () => {
      const initialState = { [REDUCER_KEY_APP]: AppStateRecord() }
      const store = configureStore(initialState, history)

      renderResult = await render(
        <Provider store={store}>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <RatesList baseCurrency={baseCurrency} />
          </IntlProvider>
        </Provider>,
      )
    })

    const { container, queryByText } = renderResult

    expect(container.firstChild).toMatchSnapshot()
    expect(container.getElementsByClassName('currency-rate').length).toEqual(
      Object.keys(testRatesResponse.rates).length,
    )
    expect(queryByText(/AUD/)).toBeInTheDocument()
    expect(queryByText(/BGN/)).toBeInTheDocument()
    expect(queryByText(/BRL/)).toBeInTheDocument()
    expect(fetcherMock).toHaveBeenCalledTimes(1)
    expect(fetcherMock).toHaveBeenCalledWith(`/latest?from=${baseCurrency}`)
  })

  it('should not render anything if nothing interesting is provided', async () => {
    fetcherMock = jest.spyOn(fetcher, 'getFranfurterData').mockImplementation(() => Promise.resolve({}))

    let renderResult: any = {}

    await act(async () => {
      renderResult = await render(<RatesList />)
    })

    const { container } = renderResult

    expect(container).toBeEmptyDOMElement()
  })
})
