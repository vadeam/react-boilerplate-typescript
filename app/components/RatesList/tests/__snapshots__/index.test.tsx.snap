// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatesList /> should render the currencies rates if loading was successful 1`] = `
<div
  class="sc-gsTEea kgaQvI"
>
  <ul
    class="sc-bdfBQB NAKmS"
  >
    <li
      class="sc-hKgJUU ggufoM"
    >
      <div
        class="sc-dlfnuX bBDutN"
      >
        <div
          class="sc-eCstlR hzQvYN currency-rate"
        >
          AUD
          :  
          1.3444
        </div>
      </div>
    </li>
    <li
      class="sc-hKgJUU ggufoM"
    >
      <div
        class="sc-dlfnuX bBDutN"
      >
        <div
          class="sc-eCstlR hzQvYN currency-rate"
        >
          BGN
          :  
          1.6474
        </div>
      </div>
    </li>
    <li
      class="sc-hKgJUU ggufoM"
    >
      <div
        class="sc-dlfnuX bBDutN"
      >
        <div
          class="sc-eCstlR hzQvYN currency-rate"
        >
          BRL
          :  
          5.1616
        </div>
      </div>
    </li>
  </ul>
</div>
`;
